src/librustc/middle/resolve_lifetime.rs

```diff
@@ -2241,19 +2242,59 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                 "this function's return type contains a borrowed value, but \
                  there is no value for it to be borrowed from"
             );
-            help!(db, "consider giving it a 'static lifetime");
+            let msg = "consider giving it a 'static lifetime";
+            match self.tcx.sess.source_map().span_to_snippet(span) {
+                Ok(ref snippet) if snippet == "&" => db.span_suggestion_with_applicability(
```

Comment:
This match has some formattings that threw me off. I think it would be easier on the eyes to have something along the lines of

```rust
Ok(ref snippet) => {
    let (sugg, app) = match &snippet[..] {
        "&" => ("&'static ".to_owned(), Applicability::MachineApplicable),
        "'_" => ("'static ".to_owned(), Applicability::MachineApplicable),
        snippet => (format!("{} + 'static", snippet), Applicability::MaybeIncorrect),
    };
    db.span_suggestion_with_applicability(span, msg, sugg, app);
}
```